source/extensions/filters/http/admission_control/admission_control.h

```diff
@@ -0,0 +1,117 @@
+
+/**
+ * Wrapper struct for admission control filter stats. @see stats_macros.h
+ */
+struct AdmissionControlStats {
+  ALL_ADMISSION_CONTROL_STATS(GENERATE_COUNTER_STRUCT)
+};
+
+using AdmissionControlProto =
+    envoy::extensions::filters::http::admission_control::v3alpha::AdmissionControl;
+
+/**
+ * Configuration for the admission control filter.
+ */
+class AdmissionControlFilterConfig {
+public:
+  AdmissionControlFilterConfig(const AdmissionControlProto& proto_config, Runtime::Loader& runtime,
+                               TimeSource&, Runtime::RandomGenerator& random, Stats::Scope& scope,
+                               ThreadLocal::SlotPtr&& tls,
+                               std::shared_ptr<ResponseEvaluator> response_evaluator);
+  virtual ~AdmissionControlFilterConfig() = default;
+
+  virtual ThreadLocalController& getController() const { NOT_IMPLEMENTED_GCOVR_EXCL_LINE; }
+
+  Runtime::RandomGenerator& random() const { return random_; }
+  bool filterEnabled() const { return admission_control_feature_.enabled(); }
+  Stats::Scope& scope() const { return scope_; }
+  double aggression() const;
+  ResponseEvaluator& responseEvaluator() const { return *response_evaluator_; }
+
+private:
+  Runtime::RandomGenerator& random_;
+  Stats::Scope& scope_;
+  const ThreadLocal::SlotPtr tls_;
+  Runtime::FeatureFlag admission_control_feature_;
+  std::unique_ptr<Runtime::Double> aggression_;
+  std::shared_ptr<ResponseEvaluator> response_evaluator_;
+};
+
+using AdmissionControlFilterConfigSharedPtr = std::shared_ptr<const AdmissionControlFilterConfig>;
+
+/**
+ * A filter that probabilistically rejects requests based on upstream success-rate.
+ */
+class AdmissionControlFilter : public Http::PassThroughFilter,
+                               Logger::Loggable<Logger::Id::filter> {
+public:
+  AdmissionControlFilter(AdmissionControlFilterConfigSharedPtr config,
+                         const std::string& stats_prefix);
+
+  // Http::StreamDecoderFilter
+  Http::FilterHeadersStatus decodeHeaders(Http::RequestHeaderMap&, bool) override;
+
+  // Http::StreamEncoderFilter
+  Http::FilterHeadersStatus encodeHeaders(Http::ResponseHeaderMap& headers,
+                                          bool end_stream) override;
+  Http::FilterTrailersStatus encodeTrailers(Http::ResponseTrailerMap& trailers) override;
+
+private:
+  static AdmissionControlStats generateStats(Stats::Scope& scope, const std::string& prefix) {
+    return {ALL_ADMISSION_CONTROL_STATS(POOL_COUNTER_PREFIX(scope, prefix))};
+  }
+
+  bool shouldRejectRequest() const;
+
+  void recordSuccess() {
+    config_->getController().recordSuccess();
+    ASSERT(deferred_record_failure_);
+    deferred_record_failure_->cancel();
+  }
+
+  void recordFailure() { deferred_record_failure_.reset(); }
+
+  AdmissionControlFilterConfigSharedPtr config_;
+  AdmissionControlStats stats_;
+  absl::optional<Cleanup> deferred_record_failure_;
```

Comment:
I don't think you want to count a downstream reset as a failure, so I think you actually probably need to understand why the stream is being reset instead of using this generic cleanup logic. Unfortunately, IIRC it's not possible for an HTTP filter to understand why a stream is being reset currently so you might need to add this. For now I think I would just delete this cleanup logic, do explicit failure tracking in responses, and take a TODO to handle upstream resets?